src/adversarial/neurify.jl

```diff
@@ -0,0 +1,297 @@
+            error("No solution, please check the problem definition.")
+        end
+
+    end
+
+    if max_violation > 0.0
+        return CounterExampleResult(:unknown), output_constraints[max_con_ind].a
+    else
+        return CounterExampleResult(:holds),   nothing
+    end
+end
+
+function constraint_refinement(solver::Neurify,
+                               nnet::Network,
+                               reach::SymbolicIntervalGradient,
+                               max_violation_con::AbstractVector{Float64},
+                               splits)
+
+    i, j, influence = get_max_nodewise_influence(solver, nnet, reach, max_violation_con, splits)
+    # We can generate three more constraints
+    # Symbolic representation of node i j is Low[i][j,:] and Up[i][j,:]
+    reach_new = forward_network(solver, Network(nnet.layers[1:i]), reach.sym.interval)
+
+    aL, bL = reach_new.sym.Low[j, 1:end-1], reach_new.sym.Low[j, end]
+    aU, bU = reach_new.sym.Up[j, 1:end-1], reach_new.sym.Up[j, end]
+
+    # custom intersection function that doesn't do constraint pruning
+    ∩ = (set, lc) -> HPolytope([constraints_list(set); lc])
+
+    subsets = [reach.sym.interval]
+
+    # If either of the normal vectors is the 0-vector, we must skip it.
+    # It cannot be used to create a halfspace constraint.
+    # NOTE: how can this come about, and does it mean anything?
+    if !iszero(aL)
+        subsets = subsets .∩ [HalfSpace(aL, -bL), HalfSpace(aL, -bL), HalfSpace(-aL, bL)]
+    end
+    if !iszero(aU)
+        subsets = subsets .∩ [HalfSpace(aU, -bU), HalfSpace(-aU, bU), HalfSpace(-aU, bU)]
+    end
+    return filter(!isempty, subsets)
+end
+
+
+function get_max_nodewise_influence(solver::Neurify,
+                                    nnet::Network,
+                                    reach::SymbolicIntervalGradient,
+                                    max_violation_con::AbstractVector{Float64},
+                                    splits)
+
+    LΛ, UΛ, radii = reach.LΛ, reach.UΛ, reach.r
+    is_ambiguous_activation(i, j) = (0 < LΛ[i][j] < 1) || (0 < UΛ[i][j] < 1)
+
+    # We want to find the node with the largest influence
+    # Influence is defined as gradient * interval width
+    # The gradient is with respect to a loss defined by the most violated constraint.
+    LG = UG = max_violation_con
+    i_max, j_max, influence_max = 0, 0, -Inf
+
+    # Backpropagation to calculate the node-wise gradient
+    for i in reverse(1:length(nnet.layers))
+        layer = nnet.layers[i]
+        if layer.activation isa ReLU
+            for j in 1:n_nodes(layer)
+                if is_ambiguous_activation(i, j)
+                    # taking `influence = max_gradient * reach.r[i][j]*k` would be
+                    # different from original paper, but can improve the split efficiency.
+                    # where `k = n-i+1`, i.e. counts up from 1 as you go back in layers.
+                    influence = max(abs(LG[j]), abs(UG[j])) * radii[i][j]
```

Comment:
@Wei-TianHao do you mind explaining this comment? I did edit it from the original wording since I changed some of the code around it, but I never understood it fully then either. It would be useful in case we want to adopt that method in the future to know why/how it is more efficient.